### PR TITLE
Encapsulate Contexts Deferred

### PR DESCRIPTION
**Pull request by anthony-murphy**
Decided it was best to break apart #4420 as i don't want to introduce unintentional errors whille merging. The end goal is still the same:

ContainerRuntime is overloaded. This PR splits out all the data store handling into it own class. There should be not functional changes here, simply moving existing functionality. Once this is out, we can work to clean up the interface. Eventually datastores could just a partitioned and virtualized DSS, and data stores could host data stores. This could be used to solve issues we see with the flat list around garbage collection, and also fulfill the dream where everything is a channel, as we could compose different container runtimes that are just a different type of dds, rather than data stores at the root. Also, hoping this will increase testabiity of the datastore code, plan to explore that in a follow up.

This PR start's that journey by encapsulating contexts deferred. As as pull in layers: requesting, creating, attaching, and summarizing we can make previous layers like this, private.

### CODE REVIEW

**Review by jatgarg — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1616,16 +1593,16 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
 
     private bindFluidDataStore(fluidDataStoreRuntime: IFluidDataStoreChannel): void {
         this.verifyNotClosed();
-        assert(this.notBoundContexts.has(fluidDataStoreRuntime.id),
+        assert(this.datastoreContexts.notBoundContexts.has(fluidDataStoreRuntime.id),
             "Store to be bound should be in not bounded set");
-        this.notBoundContexts.delete(fluidDataStoreRuntime.id);
-        const context = this.getContext(fluidDataStoreRuntime.id) as LocalFluidDataStoreContext;
+        this.datastoreContexts.notBoundContexts.delete(fluidDataStoreRuntime.id);
+        const context = this.datastoreContexts.get(fluidDataStoreRuntime.id) as LocalFluidDataStoreContext;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1605]: This asserts if it's not there... How are we so confident it's there already? And why is this the moment when we call `deferred.resolve(context)` below, as opposed to whenever the context itself was first added to the map?

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1605]: Starting to figure it out.  See my comment below in `DataStoreContexts.setupNew`.

We might still want a check up top to bail if the context isn't there at all yet, I've seen this assert get hit.

File: packages/runtime/container-runtime/src/dataStoreContexts.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public has(id: string) {
+        return this._contexts.has(id);
+    }
+
+    public setupNew(context: FluidDataStoreContext) {
+        const id = context.id;
+        assert(!this._contexts.has(id), "Creating store with existing ID");
+        this.notBoundContexts.add(id);
+        const deferred = new Deferred<FluidDataStoreContext>();
+        this.contextsDeferred.set(id, deferred);
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContexts.ts:57]: Ohhhhh-kay.  I'm starting to see the light.  `setupNew` adds the id to `notBoundContexts` and _intentionally_ doesn't resolve the deferred entry. Then `ContainerRuntime.bindFluidDataStore` removes the entry from `notBoundContexts` and resolves the deferred.  So there are actually _three_ states here.  Quite a puzzle.

[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContexts.ts:57]: So contexts added via `setupNew` need to be bound before they're available via `ContainerRuntime.getDataStore`. This seems to be the important bit.

I might play around with another approach of just using plain promises in a PromiseCache and using internal events to fire when a context is bound, and the Promise would listen to that event and resolve when the right one fires. ...that is, if I even figured this out properly.

File: packages/runtime/container-runtime/src/dataStoreContexts.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    public getDeferred(id: string): Deferred<FluidDataStoreContext> {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const deferred = this.contextsDeferred.get(id)!;
+        assert(!!deferred);
+        return deferred;
+    }
+
+    public setNew(id: string, context?: FluidDataStoreContext) {
+        assert(!!context);
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContexts.ts:77]: Never mind :P

**Review by markfields — COMMENTED** [review]
@anthony-murphy - Yay for pulling stuff out of ContainerRuntime!  I've never looked at this code, so my comments aren't about your refactor (well, 1 is), but rather the use of deferred and the obscure flow here.

Also... tests?  ;)

**Comment by markfields** [comment]
FYI I am working on a PR to do some additional light refactoring to make it clearer what's going on here, and also to fully encapsulate notBoundContexts into the new class.

**Comment by anthony-murphy** [comment]
i actually had a pr ready for that as well

**Pull request merged**